pkg/cloud/rgraph/algo/parallel_queue.go

```diff
@@ -0,0 +1,241 @@
+		workerCount: 2,
+		tracer:      func(RunInfo) {},
+	}
+	for _, o := range opts {
+		o(&cfg)
+	}
+	return &ParallelQueue[T]{
+		c:    cfg,
+		in:   make(chan struct{}, 100),
+		done: make(chan RunInfo, 100),
+	}
+}
+
+// ParallelQueue executes parallel tasks on separate goroutines.
+type ParallelQueue[T fmt.Stringer] struct {
+	c config
+
+	lock    sync.Mutex
+	state   queueState
+	pending []queueElement[T]
+	in      chan struct{}
+	done    chan RunInfo
+	active  int
+}
+
+type queueElement[T fmt.Stringer] struct {
+	ri   RunInfo
+	item T
+}
+
+type queueState int
+
+const (
+	stateNotStarted queueState = iota
+	stateRunning
+	stateDone
+)
+
+// Add an item to the queue. This method is threadsafe within op() and
+// can be called during Run(). It is NOT safe to call Add() from
+// a different, unassociated thread.
+func (q *ParallelQueue[T]) Add(item T) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	qe := queueElement[T]{
+		ri: RunInfo{
+			ID:     item.String(),
+			Queued: time.Now(),
+		},
+		item: item,
+	}
+	q.pending = append(q.pending, qe)
+	if q.state == stateRunning {
+		// Add() will always result in a call to launch() as
+		// we enqueue one element for each call to Add().
+		// This means the Run() loop will wake up AT LEAST
+		// once per `item`.
+		//
+		// `item` will in q.pending during launch() because
+		// the <-q.in will happen AFTER append(q.pending).
+		q.in <- struct{}{}
+	}
+}
+
+// Run the queue using op() to process each task. Different op()s must
+// not have interdependencies (e.g. op1 is blocked on op2 completing),
+// as this can result in a deadlock.
+//
+// Queue execution will stop if op() returns an error or the context
+// is canceled. When an error occurs, there may be goroutines that are
+// continuing to execute. Use q.WaitForOrphans() to wait for them
+// remaining goroutines.
+func (q *ParallelQueue[T]) Run(ctx context.Context, op func(context.Context, T) error) error {
+	q.lock.Lock()
+	if q.state != stateNotStarted {
+		return fmt.Errorf("Run() can only be called once (state=%d)", q.state)
+	}
+	q.state = stateRunning
+	q.lock.Unlock()
+
+	for {
+		q.lock.Lock()
+
+		klog.V(4).Infof("Run loop: pending: %d active: %d", len(q.pending), q.active)
+		if len(q.pending) == 0 && q.active == 0 {
+			q.state = stateDone
+			q.lock.Unlock()
+			return nil
+		}
+		q.launch(ctx, op)
+
+		q.lock.Unlock()
+
+		select {
+		case <-ctx.Done():
+			q.lock.Lock()
+			q.state = stateDone
+			klog.V(2).Infof("Context is Done, exiting early (pending: %d active: %d): %v", len(q.pending), q.active, ctx.Err())
+			q.lock.Unlock()
+			return ctx.Err()
+		case ri := <-q.done:
+			q.c.tracer(ri)
+			q.lock.Lock()
+			q.active--
+
+			if ri.Err != nil {
+				q.state = stateDone
+				klog.V(2).Infof("Task error, exiting early (pending: %d active: %d): %v", len(q.pending), q.active, ri.Err)
+				q.lock.Unlock()
+				return ri.Err
+			}
+
+			q.lock.Unlock()
+		case <-q.in:
+			klog.V(4).Info("<-q.in")
+			// Fall through to launch().
```

Comment:
but if the state has changed before launch() this element will not be processed because you have a check before launch(), is that the intended behavior?

Reply:
It sufficient to have it get kicked from sleep in `select`.

It's fine as the state => stateDone means we are not launching more tasks anyway.

Reply:
Let me update the comment to be more specific.